File: 454. 四数相加 II/Solution.java
class Solution {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        int ans = 0;
        Map<Integer, Integer> dict1 = new HashMap<>();
        for(int i = 0; i < nums1.length; i++){
            for(int j = 0; j < nums2.length; j++){
                int t = nums1[i] + nums2[j];
                if(dict1.containsKey(t)){
                    int c = dict1.get(t);
                    dict1.replace(t, c+1);
                }
                else{
                    dict1.put(t, 1);
                }
            }
        }

        for(int i = 0; i < nums3.length; i++){
            for(int j = 0; j < nums4.length; j++){
                int t = nums3[i] + nums4[j];
                if(dict1.containsKey(0-t)){
                    ans += dict1.get(0-t);
                }
            }
        }
        return ans;
    }
}

File: 454. 四数相加 II/Solution.py
class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        # 分别计算A+B和C+D的所有和，然后再在集合中查找需要的数和出现的次数
        ans = 0
        dict1 = {}
        for i in nums1:
            for j in nums2:
                if (i + j) in dict1.keys():
                    dict1[i+j] += 1
                else:
                    dict1[i+j] = 1

        dict2 = {}
        for i in nums3:
            for j in nums4:
                if (0 - i - j) in dict1.keys(): # 匹配成功
                    ans += dict1[0 - i - j]
        return ans
        
